File: packages/sdk/src/runtime.ts
import { NativeEngine } from '@clearings/native';
import { setImmediate as yieldTurn } from 'node:timers/promises';
import { Context, type Flow, type Spec } from './flow.js';
import { ClearingsError, copyValue, nativeError, type Value } from './values.js';
import type {
  Command,
  Event,
  Limits,
  Node,
  Plan,
  Record as CoreRecord,
  Snapshot,
  Turn,
} from './control.js';

const now = () => Math.floor(performance.now());
function integer(value: number, minimum: number, maximum: number): number {
  if (!Number.isInteger(value) || value < minimum || value > maximum)
    throw new ClearingsError('CAPACITY');
  return value;
}
function keys(value: object, allowed: string[]): void {
  if (Object.keys(value).some((k) => !allowed.includes(k))) throw new ClearingsError('UNSUPPORTED');
}
export interface RuntimeOptions {
  maxRuns?: number;
  maxNodes?: number;
  maxInFlight?: number;
  workBudget?: number;
  recordLimit?: number;
  build?: string;
}
export interface RunOptions {
  timeoutMs?: number;
  maxInFlight?: number;
  signal?: AbortSignal;
}
export interface RunRecord extends CoreRecord {
  sdk_version: string;
  adapter_versions: { id: string; version: string }[];
  adapter_versions_dropped: number;
  elapsed_ms: number;
  live_values: number;
  released_values: number;
  outcome: string;
}
interface Run {
  specs: Spec[];
  values: Map<number, Value>;
  nextValue: number;
  resolve: (value: Value) => void;
  reject: (error: ClearingsError) => void;
  removeAbort: () => void;
  deadline: number;
  started: number;
  record?: RunRecord;
  releasedValues: number;
}
export class Runtime {
  private readonly core: NativeEngine;
  private readonly limits: Limits;
  private readonly workBudget: number;
  private readonly recordLimit: number;
  private readonly build: string;
  private readonly runs = new Map<number, Run>();
  private readonly actions = new Map<number, { run: number; controller: AbortController }>();
  private events: Event[] = [];
  private pumping = false;
  private scheduled = false;
  private timer: ReturnType<typeof setTimeout> | undefined;
  private closed = false;
  private fault: ClearingsError | undefined;
  private planning = 0;
  private ownedNodes = 0;
  private history: RunRecord[] = [];
  private waiters: (() => void)[] = [];
  constructor(options: RuntimeOptions = {}) {
    keys(options, ['maxRuns', 'maxNodes', 'maxInFlight', 'workBudget', 'recordLimit', 'build']);
    this.limits = {
      max_runs: integer(options.maxRuns ?? 64, 1, 1024),
      max_nodes: integer(options.maxNodes ?? 4096, 1, 65536),
      max_in_flight: integer(options.maxInFlight ?? 32, 1, 4096),
    };
    this.workBudget = integer(options.workBudget ?? 64, 1, 4096);
    this.recordLimit = integer(options.recordLimit ?? 64, 0, 1024);
    this.build = options.build ?? 'unversioned';
    if (this.build.length > 256) throw new ClearingsError('INVALID_PLAN');
    try {
      this.core = new NativeEngine(JSON.stringify(this.limits));
    } catch (error) {
      throw nativeError(error);
    }
  }
  get records(): readonly RunRecord[] {
    return structuredClone(this.history);
  }
  snapshot(): Snapshot & { planning: number; live_values: number; owned_nodes: number } {
    const core = JSON.parse(this.core.snapshot()) as Snapshot;
    return {
      ...core,
      planning: this.planning,
      live_values: [...this.runs.values()].reduce((n, r) => n + r.values.size, 0),
      owned_nodes: this.ownedNodes,
    };
  }
  async run<I extends Value, O extends Value>(
    flow: Flow<I, O>,
    input: I,
    options: RunOptions = {},
  ): Promise<O> {
    keys(options, ['timeoutMs', 'maxInFlight', 'signal']);
    if (options.signal !== undefined && !(options.signal instanceof AbortSignal))
      throw new ClearingsError('INVALID_VALUE');
    if (this.closed) throw new ClearingsError('CLOSED');
    if (this.planning + this.runs.size >= this.limits.max_runs)
      throw new ClearingsError('CAPACITY');
    const started = performance.now();
    const deadline = now() + integer(options.timeoutMs ?? 30_000, 1, 2_147_483_647);
    const capacity = integer(
      options.maxInFlight ?? this.limits.max_in_flight,
      1,
      this.limits.max_in_flight,
    );
    const check = () => {
      if (this.closed) throw new ClearingsError('CLOSED');
      if (options.signal?.aborted) throw new ClearingsError('CANCELLED');
      if (now() >= deadline) throw new ClearingsError('TIMEOUT');
    };
    check();
    this.planning++;
    const context = new Context(flow.name, () => {
      if (this.ownedNodes >= this.limits.max_nodes) throw new ClearingsError('CAPACITY');
      this.ownedNodes++;
    });
    let reserved = 0;
    try {
      // Start the build on a host turn, and yield while compiling supported large graphs.
      await yieldTurn();
      check();
      const root = flow.build(context, copyValue(input));
      const rootId = context.reference(root);
      const reachable = new Set<number>();
      const pending = [rootId];
      let work = 0;
      while (pending.length) {
        const id = pending.pop()!;
        if (!reachable.has(id)) {
          reachable.add(id);
          pending.push(...context.nodes[id]!.deps);
        }
        if (++work % 64 === 0) {
          await yieldTurn();
          check();
        }
      }
      const selected = [...reachable].sort((a, b) => a - b);
      const mapping = new Map(selected.map((id, index) => [id, index]));
      const values = new Map<number, Value>();
      let nextValue = 1;
      const specs: Spec[] = [];
      const nodes: Node[] = [];
      for (const id of selected) {
        const original = context.nodes[id]!;
        const spec = { ...original, deps: original.deps.map((d) => mapping.get(d)!) };
        let value: number | null = null;
        if (spec.kind === 'value') {
          value = nextValue++;
          values.set(value, spec.value!);
          delete spec.value;
        }
        specs.push(spec);
        nodes.push({
          kind: spec.kind,
          deps: spec.deps,
          binding: spec.binding,
          source: spec.source,
          value,
        });
        if (nodes.length % 64 === 0) {
          await yieldTurn();
          check();
        }
      }
      check();
      const submittedAt = now();
      const plan: Plan = {
        protocol: 1,
        flow: flow.name,
        build: this.build,
        nodes,
        root: mapping.get(rootId)!,
        timeout_ms: Math.max(1, deadline - submittedAt),
        max_in_flight: capacity,
      };
      const encoded = JSON.stringify(plan);
      if (Buffer.byteLength(encoded) > 2_097_152) throw new ClearingsError('CAPACITY');
      const id = this.core.submit(encoded, submittedAt);
      reserved = selected.length;
      this.ownedNodes -= context.nodes.length - selected.length;
      const result = new Promise<Value>((resolve, reject) => {
        const abort = () => {
          this.events.push({ type: 'cancel', run: id });
          this.wake();
        };
        options.signal?.addEventListener('abort', abort, { once: true });
        this.runs.set(id, {
          specs,
          values,
          nextValue,
          resolve,
          reject,
          removeAbort: () => options.signal?.removeEventListener('abort', abort),
          deadline,
          started,
          releasedValues: 0,
        });
        if (options.signal?.aborted) abort();
      });
      context.nodes.length = 0;
      this.planning--;
      this.wake();
      return (await result) as O;
    } catch (error) {
      if (!reserved) this.ownedNodes -= context.nodes.length;
      throw nativeError(error);
    } finally {
      // A submitted run is counted in runs, so planning must finish before awaiting it.
      if (!reserved) {
        this.planning--;
        this.notifyIdle();
      }
    }
  }
  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.core.close();
    this.wake();
  }
  async drain(): Promise<void> {
    if (this.fault) throw this.fault;
    if (!this.runs.size && !this.planning) return;
    await new Promise<void>((resolve) => this.waiters.push(resolve));
    if (this.fault) throw this.fault;
  }
  private notifyIdle(): void {
    if (!this.runs.size && !this.planning) {
      const waiters = this.waiters;
      this.waiters = [];
      for (const resolve of waiters) resolve();
    }
  }
  private wake(): void {
    if (this.fault) return;
    if (this.timer) {
      clearTimeout(this.timer);
      this.timer = undefined;
    }
    if (this.pumping || this.scheduled) return;
    this.scheduled = true;
    setImmediate(() => {
      this.scheduled = false;
      void this.pump();
    });
  }
  private async pump(): Promise<void> {
    if (this.pumping) return;
    this.pumping = true;
    try {
      const events = this.events.splice(0, Math.min(this.workBudget, 128));
      const turn = JSON.parse(
        this.core.advance(JSON.stringify(events), now(), this.workBudget),
      ) as Turn;
      for (let i = 0; i < turn.commands.length; i++) {
        this.command(turn.commands[i]!);
        if ((i + 1) % 64 === 0) await yieldTurn();
      }
      this.pumping = false;
      if (turn.has_work || this.events.length) this.wake();
      else if (turn.next_wakeup_ms !== null && turn.next_wakeup_ms !== undefined)
        this.timer = setTimeout(
          () => {
            this.timer = undefined;
            this.wake();
          },
          Math.max(1, turn.next_wakeup_ms - now()),
        );
      this.notifyIdle();
    } catch (error) {
      this.pumping = false;
      this.closed = true;
      this.fault = nativeError(error);
      this.core.close();
      for (const action of this.actions.values()) action.controller.abort();
      for (const run of this.runs.values()) {
        run.removeAbort();
        run.reject(nativeError(error));
      }
      // Stop after a driver fault; drain reports the fault instead of spinning or claiming cleanup.
      for (const resolve of this.waiters) resolve();
      this.waiters = [];
    }
  }
  private command(command: Command): void {
    const run = this.runs.get(command.run);
    if (!run) throw new ClearingsError('INVALID_PLAN');
    switch (command.type) {
      case 'dispatch': {
        const spec = run.specs[command.node]!;
        if (!spec.execute || spec.binding !== command.binding)
          throw new ClearingsError('INVALID_PLAN');
        const inputs = command.inputs.map((handle) => {
          if (!run.values.has(handle)) throw new ClearingsError('INVALID_PLAN');
          return run.values.get(handle)!;
        });
        const controller = new AbortController();
        this.actions.set(command.command, { run: command.run, controller });
        const finish = (event: Event) => {
          this.actions.delete(command.command);
          this.events.push(event);
          this.wake();
        };
        Promise.resolve()
          .then(() => spec.execute!(inputs, { signal: controller.signal, deadline: run.deadline }))
          .then(
            (result) => {
              try {
                const value = copyValue(result);
                const handle = run.nextValue++;
                run.values.set(handle, value);
                finish({
                  type: 'complete',
                  run: command.run,
                  command: command.command,
                  value: handle,
                  error: null,
                });
              } catch (error) {
                finish({
                  type: 'complete',
                  run: command.run,
                  command: command.command,
                  value: null,
                  error: {
                    code: error instanceof ClearingsError ? error.code : 'INVALID_VALUE',
                    source: command.source,
                  },
                });
              }
            },
            (error) =>
              finish({
                type: 'complete',
                run: command.run,
                command: command.command,
                value: null,
                error: {
                  code:
                    error instanceof ClearingsError
                      ? error.code
                      : command.kind === 'call'
                        ? 'OPERATION_FAILED'
                        : 'TRANSFORM_FAILED',
                  source: command.source,
                },
              }),
          );
        break;
      }
      case 'release':
        run.values.delete(command.value);
        run.releasedValues++;
        if (run.record) {
          run.record.released_values = run.releasedValues;
          run.record.live_values = run.values.size;
        }
        break;
      case 'cancel':
        this.actions.get(command.command)?.controller.abort();
        break;
      case 'finished': {
        const adapters = new Map(
          run.specs
            .filter((s) => s.operation)
            .map((s) => [`${s.operation!.id}@${s.operation!.version}`, s.operation!]),
        );
        const record: RunRecord = {
          ...command.record,
          sdk_version: '0.1.0',
          adapter_versions: [...adapters.values()].slice(0, 128),
          adapter_versions_dropped: Math.max(0, adapters.size - 128),
          elapsed_ms: performance.now() - run.started,
          live_values: run.values.size,
          released_values: run.releasedValues,
          outcome: command.error?.code ?? 'SUCCEEDED',
        };
        run.record = record;
        this.history.push(record);
        if (this.history.length > this.recordLimit) this.history.shift();
        run.removeAbort();
        if (command.error)
          run.reject(new ClearingsError(command.error.code, command.error.source ?? null));
        else if (
          command.value === undefined ||
          command.value === null ||
          !run.values.has(command.value)
        )
          throw new ClearingsError('INVALID_PLAN');
        else run.resolve(run.values.get(command.value)!);
        break;
      }
      case 'dropped':
        this.ownedNodes -= run.specs.length;
        run.values.clear();
        if (run.record) run.record.live_values = 0;
        this.runs.delete(command.run);
        break;
    }
  }
}

File: python/clearings/src/clearings/runtime.py
from __future__ import annotations

import asyncio
import inspect
import json
import time
from dataclasses import dataclass, replace
from typing import Any, TypeVar, cast

from . import _control as c
from ._native import NativeEngine
from .flow import Context, Flow, OperationContext, Spec
from .values import ACTIONS, ClearingsError, Value, copy_value

InputT = TypeVar("InputT")
OutputT = TypeVar("OutputT")


def now() -> int:
    return int(time.monotonic() * 1000)


def integer(value: int, minimum: int, maximum: int) -> int:
    if type(value) is not int or not minimum <= value <= maximum:
        raise ClearingsError("CAPACITY")
    return value


def native_error(error: Exception) -> ClearingsError:
    if isinstance(error, ClearingsError):
        return error
    try:
        parsed = json.loads(str(error))
        if parsed["code"] in ACTIONS:
            return ClearingsError(parsed["code"], parsed.get("source"))
    except (ValueError, TypeError, KeyError):
        pass
    return ClearingsError("INVALID_PLAN")


@dataclass
class Run:
    specs: list[Spec]
    values: dict[int, Value]
    next_value: int
    result: asyncio.Future[Value]
    deadline: int
    started: float
    record: dict[str, Any] | None = None
    released_values: int = 0


class Runtime:
    def __init__(
        self,
        *,
        max_runs: int = 64,
        max_nodes: int = 4096,
        max_in_flight: int = 32,
        work_budget: int = 64,
        record_limit: int = 64,
        build: str = "unversioned",
    ) -> None:
        self._limits: c.Limits = {
            "max_runs": integer(max_runs, 1, 1024),
            "max_nodes": integer(max_nodes, 1, 65536),
            "max_in_flight": integer(max_in_flight, 1, 4096),
        }
        self._budget = integer(work_budget, 1, 4096)
        self._record_limit = integer(record_limit, 0, 1024)
        if len(build.encode()) > 256:
            raise ClearingsError("INVALID_PLAN")
        self._build = build
        self._core = NativeEngine(json.dumps(self._limits))
        self._loop: asyncio.AbstractEventLoop | None = None
        self._runs: dict[int, Run] = {}
        self._actions: dict[int, tuple[asyncio.Task[c.Event0], asyncio.Event]] = {}
        self._events: list[c.Event] = []
        self._pumping = False
        self._scheduled = False
        self._timer: asyncio.TimerHandle | None = None
        self._closed = False
        self._fault: ClearingsError | None = None
        self._planning = 0
        self._owned_nodes = 0
        self._history: list[dict[str, Any]] = []
        self._idle = asyncio.Event()
        self._idle.set()

    def _event_loop(self) -> asyncio.AbstractEventLoop:
        loop = asyncio.get_running_loop()
        if self._loop is not None and loop is not self._loop:
            raise ClearingsError("UNSUPPORTED")
        self._loop = loop
        return loop

    @property
    def records(self) -> list[dict[str, Any]]:
        # Records are metadata only; callers cannot mutate the runtime's stored copy.
        return cast(list[dict[str, Any]], json.loads(json.dumps(self._history)))

    def snapshot(self) -> dict[str, Any]:
        snapshot: dict[str, Any] = json.loads(self._core.snapshot())
        return {
            **snapshot,
            "planning": self._planning,
            "live_values": sum(len(r.values) for r in self._runs.values()),
            "owned_nodes": self._owned_nodes,
        }

    async def run(
        self,
        flow: Flow[InputT, OutputT],
        input: InputT,
        *,
        timeout_ms: int = 30000,
        max_in_flight: int | None = None,
    ) -> OutputT:
        loop = self._event_loop()
        if self._closed:
            raise ClearingsError("CLOSED")
        if self._planning + len(self._runs) >= self._limits["max_runs"]:
            raise ClearingsError("CAPACITY")
        started = time.monotonic()
        deadline = now() + integer(timeout_ms, 1, 2147483647)
        capacity = integer(
            self._limits["max_in_flight"] if max_in_flight is None else max_in_flight,
            1,
            self._limits["max_in_flight"],
        )

        def check() -> None:
            if self._closed:
                raise ClearingsError("CLOSED")
            if now() >= deadline:
                raise ClearingsError("TIMEOUT")

        def reserve() -> None:
            if self._owned_nodes >= self._limits["max_nodes"]:
                raise ClearingsError("CAPACITY")
            self._owned_nodes += 1

        self._planning += 1
        self._idle.clear()
        context = Context(flow.name, reserve)
        submitted = False
        try:
            await asyncio.sleep(0)
            check()
            root = flow.build(context, copy_value(input))
            root_id = context.reference(root)
            reachable: set[int] = set()
            pending = [root_id]
            work = 0
            while pending:
                node = pending.pop()
                if node not in reachable:
                    reachable.add(node)
                    pending.extend(context.nodes[node].deps)
                work += 1
                if work % 64 == 0:
                    await asyncio.sleep(0)
                    check()
            selected = sorted(reachable)
            mapping = {old: new for new, old in enumerate(selected)}
            values: dict[int, Value] = {}
            next_value = 1
            specs: list[Spec] = []
            nodes: list[c.Node] = []
            for old in selected:
                original = context.nodes[old]
                spec = replace(original, deps=[mapping[d] for d in original.deps], value=None)
                handle = None
                if spec.kind == "value":
                    handle = next_value
                    next_value += 1
                    values[handle] = original.value
                specs.append(spec)
                nodes.append(
                    {
                        "kind": spec.kind,
                        "deps": spec.deps,
                        "binding": spec.binding,
                        "source": spec.source,
                        "value": handle,
                    }
                )
                if len(nodes) % 64 == 0:
                    await asyncio.sleep(0)
                    check()
            check()
            submitted_at = now()
            plan: c.Plan = {
                "protocol": 1,
                "flow": flow.name,
                "build": self._build,
                "nodes": nodes,
                "root": mapping[root_id],
                "timeout_ms": max(1, deadline - submitted_at),
                "max_in_flight": capacity,
            }
            encoded = json.dumps(plan, ensure_ascii=False)
            if len(encoded.encode()) > 2097152:
                raise ClearingsError("CAPACITY")
            run_id = self._core.submit(encoded, submitted_at)
            result: asyncio.Future[Value] = loop.create_future()
            self._runs[run_id] = Run(specs, values, next_value, result, deadline, started)
            self._owned_nodes -= len(context.nodes) - len(selected)
            context.nodes.clear()
            submitted = True
            self._planning -= 1
            self._wake()
            try:
                return cast(OutputT, await asyncio.shield(result))
            except asyncio.CancelledError:
                self._events.append({"type": "cancel", "run": run_id})
                self._wake()
                # The public task keeps asyncio cancellation while the core settles its run.
                result.add_done_callback(
                    lambda future: None if future.cancelled() else future.exception()
                )
                raise
        except Exception as error:
            raise native_error(error) from None
        finally:
            if not submitted:
                self._owned_nodes -= len(context.nodes)
                self._planning -= 1
                self._notify_idle()

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._core.close()
        self._wake()

    async def drain(self) -> None:
        self._event_loop()
        if self._fault:
            raise self._fault
        await self._idle.wait()
        if self._fault:
            raise self._fault

    def _notify_idle(self) -> None:
        if not self._runs and not self._planning:
            self._idle.set()

    def _wake(self) -> None:
        if self._fault or self._loop is None or self._loop.is_closed():
            return
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None
        if self._pumping or self._scheduled:
            return
        self._scheduled = True
        self._loop.call_soon(self._start_pump)

    def _start_pump(self) -> None:
        self._scheduled = False
        assert self._loop is not None
        self._loop.create_task(self._pump())

    async def _pump(self) -> None:
        if self._pumping:
            return
        self._pumping = True
        try:
            count = min(self._budget, 128)
            events, self._events = self._events[:count], self._events[count:]
            turn = cast(
                c.Turn, json.loads(self._core.advance(json.dumps(events), now(), self._budget))
            )
            for index, command in enumerate(turn["commands"]):
                self._command(command)
                if (index + 1) % 64 == 0:
                    await asyncio.sleep(0)
            self._pumping = False
            if turn["has_work"] or self._events:
                self._wake()
            elif turn.get("next_wakeup_ms") is not None:
                assert self._loop is not None
                self._timer = self._loop.call_later(
                    max(0.001, (cast(int, turn["next_wakeup_ms"]) - now()) / 1000), self._wake
                )
            self._notify_idle()
        except Exception as error:
            self._pumping = False
            self._closed = True
            self._fault = native_error(error)
            self._core.close()
            for task, cancelled in self._actions.values():
                cancelled.set()
                task.cancel()
            for run in self._runs.values():
                if not run.result.done():
                    run.result.set_exception(self._fault)
            self._idle.set()

    async def _execute(
        self,
        command: c.Command0,
        run: Run,
        spec: Spec,
        inputs: list[Value],
        cancelled: asyncio.Event,
    ) -> c.Event0:
        event: c.Event0 = {
            "type": "complete",
            "run": command["run"],
            "command": command["command"],
            "value": None,
            "error": None,
        }
        try:
            assert spec.execute is not None
            outcome = spec.execute(inputs, OperationContext(cancelled, run.deadline))
            result = await outcome if inspect.isawaitable(outcome) else outcome
            value = copy_value(result)
            handle = run.next_value
            run.next_value += 1
            run.values[handle] = value
            event["value"] = handle
        except asyncio.CancelledError:
            event["error"] = {"code": "CANCELLED", "source": command["source"]}
        except Exception as error:
            event["error"] = {
                "code": error.code
                if isinstance(error, ClearingsError)
                else "OPERATION_FAILED"
                if command["kind"] == "call"
                else "TRANSFORM_FAILED",
                "source": command["source"],
            }
        return event

    def _command(self, command: c.Command) -> None:
        run = self._runs[command["run"]]
        if command["type"] == "dispatch":
            spec = run.specs[command["node"]]
            if spec.execute is None or spec.binding != command["binding"]:
                raise ClearingsError("INVALID_PLAN")
            inputs = [run.values[handle] for handle in command["inputs"]]
            cancelled = asyncio.Event()
            assert self._loop is not None
            task = self._loop.create_task(self._execute(command, run, spec, inputs, cancelled))
            self._actions[command["command"]] = (task, cancelled)

            def settled(task: asyncio.Task[c.Event0]) -> None:
                self._actions.pop(command["command"], None)
                event: c.Event0 = (
                    {
                        "type": "complete",
                        "run": command["run"],
                        "command": command["command"],
                        "value": None,
                        "error": {"code": "CANCELLED", "source": command["source"]},
                    }
                    if task.cancelled()
                    else task.result()
                )
                self._events.append(event)
                self._wake()

            task.add_done_callback(settled)
        elif command["type"] == "release":
            run.values.pop(command["value"], None)
            run.released_values += 1
            if run.record is not None:
                run.record["released_values"] = run.released_values
                run.record["live_values"] = len(run.values)
        elif command["type"] == "cancel":
            action = self._actions.get(command["command"])
            if action is not None:
                task, cancelled = action
                cancelled.set()
                task.cancel()
        elif command["type"] == "finished":
            adapters = {
                f"{s.operation['id']}@{s.operation['version']}": s.operation
                for s in run.specs
                if s.operation is not None
            }
            failure = command.get("error")
            record: dict[str, Any] = {
                **command["record"],
                "sdk_version": "0.1.0",
                "adapter_versions": list(adapters.values())[:128],
                "adapter_versions_dropped": max(0, len(adapters) - 128),
                "elapsed_ms": (time.monotonic() - run.started) * 1000,
                "live_values": len(run.values),
                "released_values": run.released_values,
                "outcome": failure["code"] if failure else "SUCCEEDED",
            }
            run.record = record
            self._history.append(record)
            if len(self._history) > self._record_limit:
                self._history.pop(0)
            if failure:
                run.result.set_exception(ClearingsError(failure["code"], failure.get("source")))
            else:
                handle = command.get("value")
                if handle is None or handle not in run.values:
                    raise ClearingsError("INVALID_PLAN")
                run.result.set_result(run.values[handle])
        elif command["type"] == "dropped":
            self._owned_nodes -= len(run.specs)
            run.values.clear()
            if run.record is not None:
                run.record["live_values"] = 0
            del self._runs[command["run"]]
